persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,680 @@
+    pure $
+        MemberBlockAttr
+            BlockAttr
+                { blockAttrDocCommentBlock = Nothing
+                , blockAttrTokens = line
+                , blockAttrPos = pos
+                }
+
+member :: Parser Member
+member = try extraBlock <|> blockAttr
+
+entityBlock :: Parser EntityBlock
+entityBlock = L.indentBlock spaceConsumerN innerParser
+  where
+    mkEntityBlock (header, members) =
+        EntityBlock
+            { entityBlockEntityHeader = header
+            , entityBlockMembers = members
+            , entityBlockDocCommentBlock = Nothing
+            }
+    innerParser = do
+        header <- entityHeader
+        pure $ L.IndentMany Nothing (return . mkEntityBlock . (header,)) member
+
+entitiesFromDocument :: Parser [EntityBlock]
+entitiesFromDocument = many entityBlock
+
+docCommentBlockText :: DocCommentBlock -> Text
+docCommentBlockText dcb = Text.unlines $ docCommentBlockLines dcb
+
+isDocComment :: CommentToken -> Bool
+isDocComment tok = case tok of
+    DocComment _ -> True
+    _ -> False
+
+docCommentBlockFromPositionedTokens
+    :: [(SourcePos, CommentToken)] -> Maybe DocCommentBlock
+docCommentBlockFromPositionedTokens ptoks =
+    case NEL.nonEmpty ptoks of
+        Nothing -> Nothing
+        Just nel ->
+            Just $
+                DocCommentBlock
+                    { docCommentBlockLines = NEL.toList $ fmap (commentContent . snd) nel
+                    , docCommentBlockPos = fst $ NEL.head nel
+                    }
+
+associateCommentLines
+    :: [(SourcePos, CommentToken)] -> [EntityBlock] -> [EntityBlock]
+associateCommentLines _ [] = []
+associateCommentLines [] es = es
+associateCommentLines cls (eh : et) =
+    applyCommentLinesToBlock candidateLines eh
+        : associateCommentLines remainingLines et
+  where
+    dcLines = dropWhile (not . isDocComment . snd) cls
+    candidateLines =
+        takeWhile
+            (\(spos, _) -> sourceLine spos < sourceLine (blockLastPos eh))
+            dcLines
+    remainingLines = drop (length candidateLines) dcLines
+
+-- | Accepts a list of (position, comment) pairs and associates them with the
+-- block and its members.
+applyCommentLinesToBlock :: (Block a) => [(SourcePos, CommentToken)] -> a -> a
+applyCommentLinesToBlock [] a = a
+applyCommentLinesToBlock cls a = blockSetDocCommentBlock dcb $ blockSetMembers commentedMembers a
+  where
+    startLine = sourceLine $ blockFirstPos a
+    headerLines = takeWhile (\(spos, _t) -> sourceLine spos < startLine) cls
+    memberLines = dropWhile (\(spos, _t) -> sourceLine spos <= startLine) cls
+    dcb = docCommentBlockFromPositionedTokens headerLines
+    commentedMembers = associateCommentLinesWithMembers memberLines (blockMembers a)
+
+associateCommentLinesWithMembers
+    :: [(SourcePos, CommentToken)] -> [Member] -> [Member]
+associateCommentLinesWithMembers [] ms = ms
+associateCommentLinesWithMembers _ [] = []
+associateCommentLinesWithMembers cls ms@(mh : mt) = do
```

Comment:
I think we can do this association much easier by using `StateT` while parsing. Briefly,

```haskell
parseThing = do
  comments <- get -- gets the current comments
  put [] -- clears the accumulated comments
  thing <- parseThingUncommented
  pure thing { thingComments = comments }
```